### DISCUSSION TITLE
[0.18] 2021-06-24

### DISCUSSION THREAD

**Discussion by darthcloud** [darthcloud/BlueRetro#46] [General]
# UX

* LED (IO17) now pulse while in inquiry mode.
* Holding BOOT SW 3 sec enable inquiry mode, short press cancel it.
* Holding BOOT SW 10 sec factory reset the adapter.
* Automatic Inquiry mode can be disabled via web config

<hr /><em>This discussion was created from the release <a href='https://github.com/darthcloud/BlueRetro/releases/tag/v0.18'>[0.18] 2021-06-24</a>.</em>

**Comment by vaderag** [comment]
Maybe a silly question but what is inquiry mode?
Thought it might be when looking for connections but I thought it was always searching for that, so not sure what it is!
Thanks!

**Comment by darthcloud** [comment]
Its controller pairing mode.

**Comment by vaderag** [comment]
Just updated to this firmware and I'm getting a solid LED on boot
(wasn't before)

FAQ says that's an error, but just wanted to confirm that it's _still_ an error in the new FW given the LED usage changes? I still seem to be able to connect to the interface
If so, do you have details of how I can check the logs?

**Reply by darthcloud** [reply]
Your LED is probably wired improperly
![led_io17](https://user-images.githubusercontent.com/3744056/123515296-d1364300-d664-11eb-8181-30579d04096c.png)

https://github.com/darthcloud/BlueRetro/wiki#getting-blueretro-debug-logs

**Reply by vaderag** [reply]
Interesting, I'm not sure if I've been looking at the wrong thing but I haven't seen that but of wiring before? I certainly haven't knowingly adjusted the led wiring i assumed it was just onboard and connected already... Should I be?

It does seem I have something wired up incorrectly however as plugging into my GameCube seems to short it (powers off but fans still run). Edit: This might be a silly question but do I need the USB plugged in or should it pull power from the console?

Edit: I was following the below which seems to have IO17 direct to pin 18 on the DB25

![Screenshot_2021-06-26-18-11-49-069_com.microsoft.emmx.jpg](https://user-images.githubusercontent.com/17175988/123520740-0749e600-d6aa-11eb-9a85-5d9f8bc1a703.jpg)

**Reply by darthcloud** [reply]
ok so you have no led, you talking about the dev kit power led.

Check this other thread:
https://github.com/darthcloud/BlueRetro/discussions/27

Keep it simple just wire the cube plug directly to esp32 (just 3 wires)

**Reply by vaderag** [reply]
> ok so you have no led, you talking about the dev kit power led.
> 
> Check this other thread:
> https://github.com/darthcloud/BlueRetro/discussions/27
> 
> Keep it simple just wire the cube plug directly to esp32 (just 3 wires)

Ah I misunderstood that then! 
Will check out that thread and maybe start again.
I did want to wire it for the DB25 for future expansion mind but given I can't see anything obviously wrong in my wiring I guess I will need to build back up...

**Reply by vaderag** [reply]
> ok so you have no led, you talking about the dev kit power led.
> 
> Check this other thread:
> https://github.com/darthcloud/BlueRetro/discussions/27
> 
> Keep it simple just wire the cube plug directly to

Followed your advice and stripped it back to basics and still having the same issue.

![IMG_20210627_104329.jpg](https://user-images.githubusercontent.com/17175988/123540021-a4e9f780-d734-11eb-9acc-6eb7e5cdfc81.jpg)

I've confirmed continuity between the wires but if I boot with or plug in after boot I get a blank screen and the machine shorts except for fans...

Any ideas?

Edit: just seen the below so adjusted the wiring (your picture filled me with simplicity visions!)
> Assembly instructions
> If using an universal FW, make sure to connect I39 to GND and IO21, IO22, IO25, IO32, IO33, I34 & I35 to GND.
> If not wiring 2nd port plug, make sure to connect IO5 (P2_D) to 3.3V
> If not wiring 3rd port plug, make sure to connect IO26 (P3_D) to 3.3V
> If not wiring 4th port plug, make sure to connect IO27 (P4_D) to 3.3V

So slightly more complex now but same issue

![IMG_20210627_110436.jpg](https://user-images.githubusercontent.com/17175988/123540555-7e798b80-d737-11eb-94d6-848714e73670.jpg)

**Reply by darthcloud** [reply]
Keep it 3 wire only, confirm your gamecube work with regular controller, then connect your esp32 to pc only and check you get the logs.

Then connect only the GND, power on the GC, does GC started fine? if so power off.

Then add 5V connection, power on GC, If GC dont start up You are using wrong pin as 5V.

**Reply by vaderag** [reply]
> Keep it 3 wire only, confirm your gamecube work with regular controller, then connect your esp32 to pc only and check you get the logs.
> 
> Then connect only the GND, power on the GC, does GC started fine? if so power off.
> 
> Then add 5V connection, power on GC, If GC dont start up You are using wrong pin as 5V.

Thank you, will try this. Original controller definitely works

I'm using universal FW so will the 3 wires only work?

I'm 99% sure that I'm connecting the correct wire from the controller as I've continuity tested everything per the diagrams. 

Edit: so just tried it and ground works fine (even ground and data) but as you suspected the 5v doesn't. As I say I'm certain I'm using the correct pin on the controller as I've continuity tested... So really confused

Edit2: just quadruple checked and 5v definitely goes to the top left. The two ground wires are connected to ground, tho I have no shield wire. The only other thing is the 3.3v which is terminated and not connected anywhere (as described). Looking at my wavebird adaptor this doesn't exist so presumably that's as expected
So no idea... Only thing I can think is either PAL cube is somehow different or I'm being an idiot somehow 🤣

**Reply by darthcloud** [reply]
When you say 3.3v "terminated" you mean unconnected?

dont plug GC 5v at all on esp32, only gnd , data, then use usb connection like wall charger for power. That work?

**Reply by darthcloud** [reply]
you probably checked but voltage between your ground and 5v wire read 5v on multimeter? Just to be 100% sure.

**Reply by vaderag** [reply]
> When you say 3.3v "terminated" you mean unconnected?
> 
> dont plug GC 5v at all on esp32, only gnd , data, then use usb connection like wall charger for power. That work?

Yeah, 3.3v is just disconnected. Sniped the wire not connected to anything

If I connect the USB to the ESP32 and boot the cube the cube boots fine (and I do seem to be able to pair my controller to the ESP32) but nothing happens (no control)

**Reply by vaderag** [reply]
> you probably checked but voltage between your ground and 5v wire read 5v on multimeter? Just to be 100% sure.

Just to confirm the question, ground and 5v on the ESP32 or ground and 5v on the cube side?

**Reply by darthcloud** [reply]
cube side with nothing connected

**Reply by darthcloud** [reply]
no control on usb power probably due to bad detection, set i39 or use gc specific firmware

**Reply by vaderag** [reply]
> no control on usb power probably due to bad detection, set i39 or use gc specific firmware

Sorry what do you mean by "set i39"

I'm tempted to go GC specific given the issues

**Reply by darthcloud** [reply]
use gc specific, i was referring to universal fw assembly instructions for GC.

**Reply by vaderag** [reply]
> use gc specific, i was referring to universal fw assembly instructions for GC.

Okay, will report back tomorrow (if I get chance!! 🤞🏻)
Thanks for your help

**Reply by vaderag** [reply]
> use gc specific, i was referring to universal fw assembly instructions for GC.

Okay, so I flashed GC specific firmware

Connected with two wires and usb, same issue, pad connects but no control

Connected with 3 wires, same issue, crash/short

Confirmed voltage at the cube side, bang on 5v where you would expect it 

Not sure what to do now! 🤷🏻‍♂️

**Reply by darthcloud** [reply]
https://github.com/darthcloud/BlueRetro/wiki#getting-blueretro-debug-logs

Get the full logs (press EN button on esp32 after power up when putty terminal ready) copy all to a text file and attach it here.

**Reply by vaderag** [reply]
[putty.txt](https://github.com/darthcloud/BlueRetro/files/6725892/putty.txt)

Attached
Thanks for your help :)

**Reply by darthcloud** [reply]
I see nothing wrong per see.

Try connect your controller in xbox mode Start+X instead.

While no controller connected, go to the advance web config (blueretro.io) and take screenshot of the mapping

**Reply by vaderag** [reply]
I'm pretty certain I was in Xbox mode anyway (Start+X) but I've just tried it again, still nothing

Screenshot:
![image](https://user-images.githubusercontent.com/17175988/123631187-5e68cb80-d80e-11eb-9a86-21ab9398196a.png)

**Reply by darthcloud** [reply]
I'm out of idea TBH, try another GC plug, change your cables and solder the cables are the only other things I can think off. Sorry.

**Reply by darthcloud** [reply]
Try a standard branded console controller, like PS4 or PS5

**Reply by vaderag** [reply]
I'll try some different controllers... in the meantime - any ideas why might be causing the 5V issue? Wondering if it's related _somehow_

Edit: Tried Switch Pro, PowerA GC, 8Bitdo SN30, Xbox One - all connect, none provide input...

The only things I can think of are:
- there is no ground from the shield - there's no wire that connects to the shield - could this be causing a problem?
- somehow related to the 5V issue... 
- my ESP32 is somehow 'different' and the IO port is different
- Having only 3 wires is causing an issue somehow (since instructions imply you need others connected or linked to 3.3v?)
- Somehow misunderstood the orientation of the adaptor diagram... (putting pic just in case...!)
![image](https://user-images.githubusercontent.com/17175988/123634508-6dea1380-d812-11eb-8f71-889b45f563b4.png)

I've used 5 different GC plugs so far (tho they are all the same brand...). If it weren't for the fact that I'd continuity tested these from every possible angle I'd be putting it down to this...

Edit2: Gonna dig out some old extension cables I was planning on keeping as extension cables and try them later...

**Reply by vaderag** [reply]
Okay, just swapped to a much older but clearly higher quality cable and I'm still having issues.
The new cable still has the issue that if I connect the 5V it won't boot, but _also_ wont boot with only the two cables connected now! :(
Only difference in my wiring is that I have a shield wire grouped in with the grounds this time :/

**Reply by vaderag** [reply]
Oh man, I think I've just realised what's going wrong - just re measured the voltages on the cube and the pictures are of the cube side, not of the connector... So my connectors are all the wrong way around (I think!)

![IMG_20210628_171124.jpg](https://user-images.githubusercontent.com/17175988/123669342-e0b7b680-d833-11eb-9d25-6e1e4490f2d3.jpg)

Which would explain the issues as 5v and 3v would be being routed incorrectly to ground...

I hope I'm right (even if I feel like a bit of an idiot!)... Will test later

**Reply by darthcloud** [reply]
Yes all connection are shown as facing the console. As noted ;)

**Reply by vaderag** [reply]
> Yes all connection are shown as facing the console. As noted ;)

Ha. I read that as the _connector_ facing the console, you know, the bit that your wiring ;)

**Reply by vaderag** [reply]
Seems to be working now (mostly)!
Button mapping is off in my PowerA Gamecube controller (which I guess I can sort through the interface), but also it seems to stop recognising input after a couple of minutes until I pull and replug (Bluetooth retains connection)

**Reply by darthcloud** [reply]
get me a log of when it happen

**Reply by vaderag** [reply]
> get me a log of when it happen

[1stRun.txt](https://github.com/darthcloud/BlueRetro/files/6732258/1stRun.txt)
[2ndRun.txt](https://github.com/darthcloud/BlueRetro/files/6732259/2ndRun.txt)

From what I can tell, it's the 
>  bt_hci_cmd_exit_periodic_inquiry

that's causing it to drop the connection
Both times it was _immediately_ after that that the controller stopped being able to control anything

EDIT: Also, just wired up the other 3 controllers and finding that when I have more than 2 the P3 and P4 controllers get 'stuck' with massive input lag... will try grab a log of this later...

**Reply by darthcloud** [reply]
this is only problem for the powerA? This should not happen with the others controller right?

**Reply by vaderag** [reply]
The drop in connection does seem to be related only to the PowerA (as far as I can tell)
The lag/stop is on all controllers (Switch Pro, Xbox One S tested)

**Reply by darthcloud** [reply]
For PowerA since all your controller are now associated with BlueRetro, you could now disable inquiry mode:
![chrome_Yj0jpm1S82](https://user-images.githubusercontent.com/3744056/123783087-fc5aa580-d8a3-11eb-8af4-937e1bea620c.png)
Set it to manual and click save. Power cycle BlueRetro/Console.

For lag I can't do much, for 4+ player (if mixing controller type) it's important for all controller to have a report interval <= 8ms. Mixing PS4, PS5 and Xbox One S (1st party model) usually work well. Also using all identical controller should work good too.

**Reply by vaderag** [reply]
> For PowerA since all your controller are now associated with BlueRetro, you could now disable inquiry mode:
> ![chrome_Yj0jpm1S82](https://user-images.githubusercontent.com/3744056/123783087-fc5aa580-d8a3-11eb-8af4-937e1bea620c.png)
> Set it to manual and click save. Power cycle BlueRetro/Console.
> 
> For lag I can't do much, for 4+ player (if mixing controller type) it's important for all controller to have a report interval <= 8ms. Mixing PS4, PS5 and Xbox One S (1st party model) usually work well. Also using all identical controller should work good too.

Will give that a try.
Presumably with the manual setting is just a 3s press of Boot and can pair a new controller?

Just to confirm, assuming the powerA works after that, if I had 4 power a (or 4 8bit do) then that should work? Would they need to be identical? I.e. 4 8bitdo SN30 or would variants be okay? Like could use the Gbros for original GC to Bluetooth with an SN30?

**Reply by darthcloud** [reply]
I can't tell for sure, it's RF so it's very dependent on how crowded is your 2.4GHz environment, downtown apartment with a lot of neighboring Bluetooth, WI-Fi, surveillance camera, etc will likely perform bad at 4P vs a suburb house environment.

Mixing PS4, PS5 and Xbox One S work well in my personal experience. WiiU controller always lag at 3P+. But I can't tell for other mixing as I haven't tested that. In theory identical model time should perform well, but I haven't tested that my self.

**Reply by vaderag** [reply]
Okay, thanks. I got the impression it was more due to memory than interference but I can see how that could just be my perception - perhaps they were interfering after all. That said, if they were interfering wouldn't that affect all controllers at the same time? It seemed that 1 and 2 were working okay still but 3/4 we sticking/lagging?

One (final?!) question/thought - I found that all the controllers in default mapping were incorrect (triggers not working, a/b swapped etc), so will be updating them all (suggestion that the defaults might want updating)
Do you need to have the device connected to change the mapping? If not, how do you know which device is "Device 1" etc?

**Reply by darthcloud** [reply]
Read the system specific manual:
https://github.com/darthcloud/BlueRetro/wiki/BlueRetro-System-Specific-User-Manual#gamecube

And check default mapping:
https://docs.google.com/spreadsheets/d/e/2PACX-1vRln_dhkahEIhq4FQY_p461r5qvLn-Hkl89ZtfyIOGAqdnPtQZ5Ihfsjvd94fRbaHX8wU3F-r2ODYbM/pubhtml

Mapping is base on button location, not specific controller names. Default mapping wont change, as far as I'm concerned the default for swith pro, ps3, ps4, ps5, xbox match the location on the original GC pad.

**Reply by darthcloud** [reply]
To access web config, its required to not have any device connected. Device 1 is the first one connect, dev 2 the 2nd, etc

**Reply by vaderag** [reply]
> To access web config, its required to not have any device connected. Device 1 is the first one connect, dev 2 the 2nd, etc

Ah, so if you're using different controllers then you need to ensure you always connect them in the same order then
Would there be any way to store specific device settings so they could be connected in any order? (guessing this is a feature request rather than something that can just be done)

**Reply by darthcloud** [reply]
This would add a lot of complexity for little benefits. Not going to happen sorry.

**Reply by vaderag** [reply]
> This would add a lot of complexity for little benefits. Not going to happen sorry.

Couldn't you link the mapping to the bluetooth id rather than device connection order?

**Reply by vaderag** [reply]
> > This would add a lot of complexity for little benefits. Not going to happen sorry.
> 
> Couldn't you link the mapping to the bluetooth id rather than device connection order?

Also just to explain benefits - would like to use PowerA Gamecube controller for gamecube (as player one), generally. I emulate SNES on my Gamecube - I wish to use SN30 for this, and there are games more suited to SNES controller (e.g. Zelda Four Swords originally designed for GBA control)
Mapping will be different for both controllers, yet no way to do this

This is my gamecube use case, but same applies on other consoles

Decided it's worth creating a new idea on this for discussion: https://github.com/darthcloud/BlueRetro/discussions/50

**Comment by Micha213** [comment]
Today I updated from 0.16 to 0.19.1.
I was able to test the function of the LED on IO17 and the new function of the BOOT button.
I am writing in this thread because the function was added with v0.18.

As I have seen, there are two settings for inquiry mode: Auto and Manuel.
Would it be possible to add a third option?
Timeout: The inquiry mode is active 30 or 60 seconds after the start (or reset) and then goes out by itself. So that the LED doesn't blink all the time.

Another thing: would it also be possible to display the current FW version via the web interface?

**Reply by darthcloud** [reply]
What the rational of that third option? You run the adapter without connecting controller to it?

**Reply by Micha213** [reply]
Ah, I just see that there is already a timeout there. I hadn't seen that at first because the LED flashes for a long time before it goes out.
I should test more carefully before I post ideas, sorry.
Possibly you could end the inquiry mode sooner after a controller is connected. But that's just my personal preference, that's okay too.
Sorry for the misunderstanding.

**Reply by darthcloud** [reply]
No worries, feel free to suggest any idea. 👍 

it's around 60sec right now IIRC. It's indeed long, I'll think about lowering to 30sec.

**Reply by darthcloud** [reply]
@Micha213 [v1.0-beta](https://github.com/darthcloud/BlueRetro/releases/tag/v1.0-beta)  is out with your suggestions implemented. I reduced timeout to ~10sec and OTA page logs will print firmware version if current version loaded is => v1.0-beta
